### PR TITLE
[jaeger] Added Kafka provisioning

### PR DESCRIPTION
**Pull request by arpitjindal97**

### CODE REVIEW

**Comment by naseemkullah** [comment]
Thanks @arpitjindal97, please go ahead and include the deployment of ingester here since provisioning kafka is part of that larger scoped task of deploying with the ingester component setup.

**Comment by arpitjindal97** [comment]
Resolves #38 
Resolves #27

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 1]
```diff
@@ -0,0 +1,147 @@
+{{- if .Values.ingester.enabled -}}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "jaeger.fullname" . }}-ingester
+  labels:
+    app.kubernetes.io/name: {{ include "jaeger.name" . }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
    {{- include "jaeger.labels" . | nindent 4 }}
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    helm.sh/chart: {{ include "jaeger.chart" . }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    app.kubernetes.io/component: ingester
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    app.kubernetes.io/instance: {{ .Release.Name }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+{{- end }}
+spec:
+{{- if not .Values.ingester.autoscaling.enabled }}
+  replicas: {{ .Values.ingester.replicaCount }}
+{{- end }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "jaeger.name" . }}
+      app.kubernetes.io/component: ingester
+      app.kubernetes.io/instance: {{ .Release.Name }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  annotations:
+{{ toYaml .Values.ingester.annotations | indent 4 }}
+{{- end }}
+spec:
+{{- if not .Values.ingester.autoscaling.enabled }}
+  replicas: {{ .Values.ingester.replicaCount }}
+{{- end }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "jaeger.name" . }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
      {- include "jaeger.selectorLabels" . | nindent 6 }}
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  strategy:
+    type: Recreate
+  template:
+    metadata:
+      annotations:
+{{- if .Values.ingester.podAnnotations }}
+{{ toYaml .Values.ingester.podAnnotations | indent 8 }}
+{{- end }}
+      labels:
+        app.kubernetes.io/name: {{ include "jaeger.name" . }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
        {{- include "jaeger.selectorLabels" . | nindent 8 }}
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        app.kubernetes.io/component: ingester
+        app.kubernetes.io/instance: {{ .Release.Name }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        - name: {{ .name }}
+          configMap:
+            name: {{ .configMap }}
+      {{- end }}
+      {{- range .Values.ingester.extraSecretMounts }}
+        - name: {{ .name }}
+          secret:
+            secretName: {{ .secretName }}
+      {{- end }}
+
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+{{- end -}}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
{{- end -}}

```

File: charts/jaeger/templates/_helpers.tpl [hunk 11]
```diff
@@ -217,3 +241,8 @@ Configure list of IP CIDRs allowed access to load balancer (if supported)
   {{- end }}
 {{- end }}
 {{- end -}}
+
+{{- define "helm-toolkit.utils.joinListWithComma" -}}
+{{- $local := dict "first" true -}}
+{{- range $k, $v := . -}}{{- if not $local.first -}},{{- end -}}{{- $v -}}{{- $_ := set $local "first" false -}}{{- end -}}
+{{- end -}}
```
Comment (naseemkullah) [inline charts/jaeger/templates/_helpers.tpl]:
```suggestion
{{- end -}}

```

File: charts/jaeger/templates/ingester-deploy.yaml [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        imagePullPolicy: {{ .Values.ingester.pullPolicy }}
+        env:
+          {{- range $key, $value := .Values.ingester.cmdlineParams }}
+          - name: {{ $key | replace "." "_" | replace "-" "_" | upper | quote }}
+            value: {{ $value | quote }}
+          {{- end }}
+          - name: SPAN_STORAGE_TYPE
+            value: {{ .Values.storage.type }}
+          - name: KAFKA_CONSUMER_BROKERS
+            value: {{ include "helm-toolkit.utils.joinListWithComma" .Values.storage.kafka.broker }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/ingester-deploy.yaml]:
```suggestion
            value: {{ include "helm-toolkit.utils.joinListWithComma" .Values.storage.kafka.brokers }}
```

File: charts/jaeger/templates/collector-deploy.yaml [hunk 13]
```diff
@@ -48,6 +48,14 @@ spec:
           - name: COLLECTOR_ZIPKIN_HTTP_PORT
             value: {{ .Values.collector.service.zipkinPort | quote }}
           {{- end }}
+          {{- if .Values.ingester.enabled }}
+          - name: SPAN_STORAGE_TYPE
+            value: kafka
+          - name: KAFKA_PRODUCER_BROKERS
+            value: {{ include "helm-toolkit.utils.joinListWithComma" .Values.storage.kafka.broker }}
```
Comment (naseemkullah) [inline charts/jaeger/templates/collector-deploy.yaml]:
```suggestion
            value: {{ include "helm-toolkit.utils.joinListWithComma" .Values.storage.kafka.brokers }}
```

File: charts/jaeger/README.md [hunk 14]
```diff
@@ -222,12 +222,28 @@ The following table lists the configurable parameters of the Jaeger chart and th
 | `collector.extraConfigmapMounts` | Additional collector configMap mounts | `[]` |
 | `collector.extraSecretMounts` | Additional collector secret mounts | `[]` |
 | `collector.samplingConfig` | [Sampling strategies json file](https://www.jaegertracing.io/docs/latest/sampling/#collector-sampling-configuration) | `nil` |
+| `ingester.autoscaling.enabled` | Enable horizontal pod autoscaling | `false` |
```
Comment (naseemkullah) [inline charts/jaeger/README.md]:
```suggestion
| `ingester.enabled` | Enable ingester component, collectors will write to Kafka | `false` |
| `ingester.autoscaling.enabled` | Enable horizontal pod autoscaling | `false` |
```

Comment (naseemkullah) [reply charts/jaeger/README.md]: Adding the `ingester.enabled` value here @arpitjindal97

File: charts/jaeger/Chart.yaml [hunk 15]
```diff
@@ -2,7 +2,7 @@ apiVersion: v1
 appVersion: 1.16.0
 description: A Jaeger Helm chart for Kubernetes
 name: jaeger
-version: 0.18.5
+version: 0.18.6
```
Comment (naseemkullah) [inline charts/jaeger/Chart.yaml]: nit: but this adds a new feature so a minor bump would be appropriate

Comment (arpitjindal97) [reply charts/jaeger/Chart.yaml]: bumped to 0.19.0

**Comment by naseemkullah** [comment]
lgtm thanks @arpitjindal97

**Pull request merged**